Tooltip docs (#1098)
adobe/react-spectrum: packages/@react-aria/tooltip/docs/useTooltipTrigger.mdx

of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. -->

import {Layout} from '@react-spectrum/docs';
export default Layout;

import docs from 'docs:@react-aria/tooltip';
import statelyDocs from 'docs:@react-stately/tooltip';
import {HeaderInfo, FunctionAPI, TypeContext, InterfaceType, TypeLink} from '@react-spectrum/docs';
import packageData from '@react-aria/switch/package.json';

```jsx import
import {useTooltipTrigger} from '@react-aria/tooltip';
import {useTooltip} from '@react-aria/tooltip';
```

---
category: Overlays
keywords: [tooltip, aria]
---

# useTooltipTrigger

<p>{docs.exports.useTooltipTrigger.description}</p>

<HeaderInfo
  packageData={packageData}
  componentNames={['useTooltipTrigger']}
  sourceData={[
    {type: 'W3C', url: 'https://www.w3.org/TR/wai-aria-1.2/#tooltip'}
  ]} />

## API

> <FunctionAPI function={docs.exports.useTooltipTrigger} links={docs.links} />

Reviewer: How do you feel about adding the the api for `useTooltip` here as well?
```suggestion
<FunctionAPI function={docs.exports.useTooltipTrigger} links={docs.links} />
<FunctionAPI function={docs.exports.useTooltip} links={docs.links} />
```

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. -->

import {Layout} from '@react-spectrum/docs';
export default Layout;

import docs from 'docs:@react-aria/tooltip';
import statelyDocs from 'docs:@react-stately/tooltip';
import {HeaderInfo, FunctionAPI, TypeContext, InterfaceType, TypeLink} from '@react-spectrum/docs';
import packageData from '@react-aria/switch/package.json';

```jsx import
import {useTooltipTrigger} from '@react-aria/tooltip';
import {useTooltip} from '@react-aria/tooltip';
```

---
category: Overlays
keywords: [tooltip, aria]
---

# useTooltipTrigger

<p>{docs.exports.useTooltipTrigger.description}</p>

<HeaderInfo
  packageData={packageData}
  componentNames={['useTooltipTrigger']}
  sourceData={[
    {type: 'W3C', url: 'https://www.w3.org/TR/wai-aria-1.2/#tooltip'}
  ]} />

## API

<FunctionAPI function={docs.exports.useTooltipTrigger} links={docs.links} />

## Features

> There is no native HTML element for tooltips. There are many guides out there for
> tooltips, however, they are not easy to make accessible, especially the CSS only ones.

Reviewer 2: You could argue the `title` attribute could be used for this...

```suggestion
The HTML `title` attribute can be used to create a tooltip, but it cannot be styled. Custom styled 
tooltips can be hard to build in an accessible way. `useTooltipTrigger` and `useTooltip` help build 
fully accessible tooltips that can be styled as needed.
```

Author: good point, i'd forgotten about title since there's almost nothing you can do with it

* Keyboard focus management and cross browser normalization
* Hover management and cross browser normalization
* Labeling support for screen readers (aria-describedby)

> * Exposed as a tooltip to assistive technology via ARIA

Reviewer 2: Maybe add in the warmup/cooldown features?

## Anatomy

> A tooltip consists of two parts, the target trigger and the tooltip itself.
> Users may reveal the tooltip by hovering or focusing the target trigger.

Reviewer 2: target trigger is a bit redundant?

```suggestion
A tooltip consists of two parts, the trigger element and the tooltip itself.
Users may reveal the tooltip by hovering or focusing the trigger.
```

`useTooltipTrigger` returns props to be spread onto its target trigger and the tooltip:

<TypeContext.Provider value={docs.links}>
  <InterfaceType properties={docs.links[docs.exports.useTooltipTrigger.return.id].properties} />

> </TypeContext.Provider>

Reviewer 2: Also include this for `useTooltip`?

Tooltip state is managed by the <TypeLink links={statelyDocs.links} type={statelyDocs.exports.useTooltipTriggerState} />
hook in `@react-stately/tooltip`. The state object should be passed as an option to `useTooltipTrigger`.

## Example

> This example uses a Tooltip that renders in an absolute position next to its target. We need to supply it with the correct
> aria attributes, which we can get from the <TypeLink links={docs.links} type={docs.exports.useTooltip} /> hook also in `@react-aria/tooltip`.

Reviewer 2: ```suggestion
This example implements a Tooltip that renders in an absolute position next to its target. The <TypeLink links={docs.links} type={docs.exports.useTooltip} /> hook applies the correct ARIA attributes to the tooltip, and `useTooltipTrigger` associates it to the trigger element.
```

```tsx example
import {useTooltipTriggerState} from '@react-stately/tooltip';
import {
  useOverlay,

>   useOverlayPosition,

Reviewer 2: Unused?

} from '@react-aria/overlays';
import {mergeProps} from '@react-aria/utils';

function Tooltip(props) {
  let {tooltipProps} = useTooltip(props);

  return (
    <span
      style={{
        position: 'absolute',
        right: '-10px',
        top: '0',
        transform: 'translateX(100%)',
        backgroundColor: 'white',
        color: 'black',
        padding: '5px'
      }}
      {...mergeProps(props, tooltipProps)} >
      {props.children}
    </span>
  );
}

> function TooltipedButton(props) {

Reviewer 2: kinda strange name :D maybe just `Example` or `TooltipExample` since it's not really a component

  let state = useTooltipTriggerState(props);
  let ref = React.useRef();
  // Handle interacting outside the dialog and pressing
  // the Escape key to close the modal.

>   let {overlayProps} = useOverlay(

Reviewer 2: This isn't needed I don't think? React Spectrum's TooltipTrigger doesn't use it?

Author: yep, not needed, i'd had an issue where i couldn't get it to disappear on Esc and forgot to remove this after resolving it

    props,
    ref
  );

  // Get props for the trigger and its tooltip
  let {triggerProps, tooltipProps} = useTooltipTrigger(props, state, ref);

  return (
    <span style={{position: 'relative'}}>

>         <button ref={ref} {...triggerProps}>I have a tooltip</button>

Reviewer 2: 2 space indent?

import packageData from '@react-aria/tooltip/package.json';

```jsx import
import {useTooltipTrigger} from '@react-aria/tooltip';
import {useTooltip} from '@react-aria/tooltip';
```

---
category: Overlays
keywords: [tooltip, aria]
---

# useTooltipTrigger

<p>{docs.exports.useTooltipTrigger.description}</p>

<HeaderInfo
  packageData={packageData}
  componentNames={['useTooltipTrigger']}
  sourceData={[
    {type: 'W3C', url: 'https://www.w3.org/TR/wai-aria-1.2/#tooltip'}
  ]} />

## API

`useTooltipTrigger`
<FunctionAPI function={docs.exports.useTooltipTrigger} links={docs.links} />

`useTooltip`
<FunctionAPI function={docs.exports.useTooltip} links={docs.links} />

## Features

The HTML `title` attribute can be used to create a tooltip, but it cannot be styled. Custom styled
tooltips can be hard to build in an accessible way. `useTooltipTrigger` and `useTooltip` help build
fully accessible tooltips that can be styled as needed.

* Keyboard focus management and cross browser normalization
* Hover management and cross browser normalization
* Labeling support for screen readers (aria-describedby)
* Exposed as a tooltip to assistive technology via ARIA

> * Global warmup/cooldown timers

Reviewer 2: ```suggestion
* Matches native tooltip behavior with delay on hover of first tooltip and no delay on subsequent tooltips.
```
